girder/models/model_base.py

```diff
@@ -569,32 +569,103 @@ def subtreeCount(self, doc):
-        :type fields: `str, list of strings or tuple of strings for fields to be included from the
-            document, or dict for an inclusion or exclusion projection`.
-        """
-        if fields is None:
-            return False
+def _isInclusionProjection(fields):
+    """
+    Test whether a projection filter is an inclusion filter (whitelist) or exclusion
+    projection (blacklist) of fields, as defined by MongoDB find() method `projection` param.
+
+    :param fields: A mask for filtering result documents by key, or None to return the full
+        document, passed to MongoDB find() as the `projection` param.
+    :type fields: `str, list of strings or tuple of strings for fields to be included from the
+        document, or dict for an inclusion or exclusion projection`.
+    """
+    if fields is None:
+        return False
+
+    if not isinstance(fields, dict):
+        # If this is a list/tuple/set, that means inclusion
+        return True
+
+    for k, v in six.viewitems(fields):
+        if k != '_id':
+            # We are only allowed either inclusion or exclusion keys in a dict, there can be no
+            # mixing of these, with the only exception being that the `_id` key can be set as
+            # an exclusion field in a dict that otherwise holds inclusion fields.
+            return v
 
+    # Empty dict or just _id field
+    return fields.get('_id', True)
+
+
+def _overwriteFields(fields, overwrite):
+    """
+    Overwrite the projection filter to either include (in the case of an inclusion filter) or
+    not exclude (in the case of an exclusion filter) the contents of overwrite.
+
+    :param fields: A mask for filtering result documents by key, or None to return the full
+        document, passed to MongoDB find() as the `projection` param.
+    :type fields: `str, list of strings, or tuple of strings for fields to be included from the
+        document, or dict for an inclusion or exclusion projection`.
+    :param overwrite: Additional document key(s) to be included or not excluded in fields.
+    :type overwrite: `str, list of strings, or tuple of strings for fields to be included from the
+        document.`
+    """
+    if fields is None:
+        return fields
+
+    overwrite = list(overwrite)
+    if _isInclusionProjection(fields):
         if not isinstance(fields, dict):
-            # If this is a list/tuple/set, that means inclusion
-            return True
+            # Inclusion projection (str, list, or tuple)
+            copy = list(fields)
+            for entry in overwrite:
+                if entry not in copy:
+                    copy.append(entry)
+        else:
+            # Inclusion projection (dict)
+            copy = dict(fields)
+            for entry in overwrite:
+                copy[entry] = True
+    else:
+        # Exclusion projection (dict)
+        copy = dict(fields)
+        for entry in overwrite:
+            if entry in copy.keys():
```

Comment:
If you're going to check if something is in a `dict`, you should definitely just do:
```
if entry in copy:
```
otherwise, you're creating an entire list / view of the keys every time through the loop, then doing a list search, which itself is linear-time.

Alternatively, you could just omit the check entirely, and do:
```
for entry in overwrite:
    copy.pop(entry, None)
```

Reply:
This would be a great behavior for `dict.__sub__` IMO, but sadly that operator is not implemented.